tests/test_conformer_rel_pos.py

```diff
@@ -0,0 +1,206 @@
+            input_dim=input_dim,
+            hidden_dim=hidden_dim,
+            dropout=dropout,
+            activation=activation,
+            dropout_broadcast_axes=dropout_broadcast_axes,
+        )
+        conf_ffn_part = ConformerPositionwiseFeedForwardV2(cfg)
+        y = conf_ffn_part(x)
+        return y.shape
+
+    for input_dim, hidden_dim, dropout, activation, dropout_broadcast_axes in product(
+        [10, 20], [100, 200], [0.1, 0.3], [nn.functional.silu, nn.functional.relu], [None, "B", "T", "BT"]
+    ):
+        input_shape = (10, 100, input_dim)
+        assert get_output_shape(input_shape, input_dim, hidden_dim, dropout, activation) == input_shape
+
+
+def test_ConformerMHSARelPosV1():
+    def get_output_shape(
+        input_shape,
+        seq_len,
+        input_dim,
+        with_bias=True,
+        num_att_heads=8,
+        att_weights_dropout=0.1,
+        dropout=0.1,
+        learnable_pos_emb=True,
+        with_linear_pos=False,
+        separate_pos_emb_per_head=False,
+        rel_pos_clip=16,
+        with_pos_bias=False,
+        pos_emb_dropout=0.0,
+        dropout_broadcast_axes=None,
+    ):
+        assert len(input_shape) == 3 and input_shape[-1] == input_dim
+
+        cfg = ConformerMHSARelPosV1Config(
+            input_dim=input_dim,
+            num_att_heads=num_att_heads,
+            with_bias=with_bias,
+            att_weights_dropout=att_weights_dropout,
+            dropout=dropout,
+            learnable_pos_emb=learnable_pos_emb,
+            with_linear_pos=with_linear_pos,
+            separate_pos_emb_per_head=separate_pos_emb_per_head,
+            rel_pos_clip=rel_pos_clip,
+            with_pos_bias=with_pos_bias,
+            pos_emb_dropout=pos_emb_dropout,
+            dropout_broadcast_axes=dropout_broadcast_axes,
+        )
+        conf_mhsa_rel_pos = ConformerMHSARelPosV1(cfg)
+        input_tensor = torch.randn(input_shape)
+        sequence_mask = torch.less(torch.arange(input_shape[1])[None, :], torch.tensor(seq_len)[:, None])
+
+        output = conf_mhsa_rel_pos(input_tensor, sequence_mask)
+
+        return list(output.shape)
+
+    # with key padding mask
+    input_shape = [4, 15, 32]  # B,T,F
+    seq_len = [15, 12, 10, 15]
+
+    for learnable_pos_emb, with_pos_bias, pos_emb_dropout, with_linear_pos, separate_pos_emb_per_head in product(
+        [True, False], [True, False], [0.0, 0.1], [True, False], [True, False]
+    ):
+        assert get_output_shape(
+            input_shape,
+            seq_len,
+            32,
+            learnable_pos_emb=learnable_pos_emb,
+            with_pos_bias=with_pos_bias,
+            pos_emb_dropout=pos_emb_dropout,
+            with_linear_pos=with_linear_pos,
+            separate_pos_emb_per_head=separate_pos_emb_per_head,
+        ) == [4, 15, 32]
+
+
+def test_ConformerMHSARelPosV1_against_Espnet():
+    from espnet2.asr_transducer.encoder.modules.attention import RelPositionMultiHeadedAttention
```

Comment:
```suggestion
        from espnet.nets.pytorch_backend.transformer.attention import RelPositionMultiHeadedAttention

```

Reply:
these two modules are unfortunately not identical, e.g. in the rel_shift function and the mask conversion logic.

Reply:
So you say the test does not work then?